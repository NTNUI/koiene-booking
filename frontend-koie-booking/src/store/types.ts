// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
