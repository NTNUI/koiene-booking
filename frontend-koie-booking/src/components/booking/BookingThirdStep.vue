<template>
  <v-container :class="$style.container">
    <v-row>
      <v-col class="global-card-outlined">
        <v-card class="global-card-outlined">
          <h3 :class="$style.priceStyle">Total price: {{ price }}kr</h3>
          <Card
            :class="[{ complete }, $style.stripeCard]"
            :options="stripeOptions"
            stripe="pk_test_TagT9jGDj7CN9NOQfTnueTxz"
            @change="complete = $event.complete"
          />
          <v-btn
            color="blue"
            class="pay-with-stripe"
            :disabled="!complete"
            width="100%"
            :class="$style.payCardBtn"
            @click="pay"
            >Pay with credit card</v-btn
          >
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue';
import { Card, createToken } from 'vue-stripe-elements-plus';
import { stripeKey, stripeOptions } from '@/stripeConfig.json';
import { BookingThirdStepData } from '@/types/booking';

export default Vue.extend({
  components: { Card },
  data(): BookingThirdStepData {
    return {
      complete: false,
      stripeOptions: {},
      totalPrice: 0
    };
  },
  computed: {
    price(): number {
      return this.$store.state.booking.bookingData.booking.price || 0;
    }
  },
  mounted() {
    if (this.$store.state.booking.error) {
      this.$store.dispatch('booking/SET_STEP', 1);
      this.$notify({
        group: 'memberships_system',
        type: 'error',
        duration: 3000,
        text: `${this.$t('booking.error_msg')}`
      });
    }
  },
  methods: {
    pay() {
      createToken().then(async (data: any) => {
        const transaction_id = await this.$store.state.booking.bookingData.booking.booking_transaction_id;
        const values = { token: data.token.id, id: transaction_id };
        this.$store.dispatch('booking/INITIATE_PAYMENT', values);
      });
    }
  }
});
</script>

<style lang="scss" module>
.container {
  width: 400px;
}

.stripeCard {
  width: 100%;
  border: 1px solid grey;
  padding: 10px;
  background-color: white;
  border-radius: 4px;
}

.payCardBtn {
  margin-top: 10px !important;
  width: 100% !important;
}

.stripeCard.complete {
  border: 2px solid rgb(29, 252, 29);
}
.priceStyle {
  margin-bottom: 20px;
  margin-top: 5px;
  text-align: center;
}
</style>
