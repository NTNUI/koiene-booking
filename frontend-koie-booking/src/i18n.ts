import Vue from 'vue';
import VueI18n, { LocaleMessages } from 'vue-i18n';
import enJson from './locales/en.json';

Vue.use(VueI18n);

function loadLocaleMessages(): LocaleMessages {
  // Testing currently only supports manually importing of json-files. English is the only language that gets loaded now
  let locales: any;
  const messages: LocaleMessages = {};
  if (process.env.NODE_ENV === 'test') {
    messages['en'] = enJson;
  } else {
    locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
    locales.keys().forEach((key: any) => {
      const matched = key.match(/([A-Za-z0-9-_]+)\./i);
      if (matched && matched.length > 1) {
        const locale = matched[1];
        messages[locale] = locales(key);
      }
    });
  }
  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages()
});
