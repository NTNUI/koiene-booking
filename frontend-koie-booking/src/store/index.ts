import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
