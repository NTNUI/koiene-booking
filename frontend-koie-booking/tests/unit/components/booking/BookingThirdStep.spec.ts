import Vue from 'vue';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
import store from '@/store';
import i18n from '@/i18n';
import scssVars from '@/styles/variables.scss';
import flushPromises from 'flush-promises';
import mockAxios from 'jest-mock-axios';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue } from '@vue/test-utils';

// Components or views
import BookingThirdStep from '@/components/booking/BookingThirdStep.vue';

describe('Component BookingThirdStep.vue', () => {
  // Router not needed for this test-suite
  let wrapper: any;
  let localVue: any;
  let vuetify: any;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(Vuetify);
    localVue.use(Vuex);
    localVue.prototype.$scssVars = scssVars;
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it('Matches snapshot', async () => {
    wrapper = mount(BookingThirdStep, {
      localVue,
      vuetify,
      i18n,
      store,
      computed: {
        price() {
          return 120;
        },
      },
      stubs: ['Card'],
    });

    expect(wrapper).toMatchSnapshot();
  });
});
