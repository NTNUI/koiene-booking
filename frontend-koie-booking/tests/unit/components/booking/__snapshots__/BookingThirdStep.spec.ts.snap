// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component BookingThirdStep.vue Matches snapshot 1`] = `
<div class="container container">
  <div class="row">
    <div class="global-card-outlined col">
      <div class="global-card-outlined v-card v-sheet theme--light">
        <h3 class="priceStyle">Total price: 120kr</h3>
        <card-stub stripe="pk_test_TagT9jGDj7CN9NOQfTnueTxz" options="[object Object]" class="stripeCard"></card-stub> <button type="button" disabled="disabled" class="pay-with-stripe v-btn v-btn--contained v-btn--disabled theme--light v-size--default payCardBtn" style="width: 100%;"><span class="v-btn__content">Pay with credit card</span></button>
      </div>
    </div>
  </div>
</div>
`;
